src/types/web-resource.ts

```diff
@@ -0,0 +1,108 @@
+import * as TypeUtils from '../type-utils';
+import { storageRegistry } from '../storage-adapters/index';
+
+/**
+ * This is how WebResources are stored on the DB
+ */
+export type WebResourceRef = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+/**
+ * This is how WebResources are defined on creation
+ */
+export type WebResourceInput = {
+	filename: string;
+	data: Buffer;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+	storage: string;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
```

Comment:
Sorry for the late thought on this. Modelling the AtomicType WebResource as an underlying complext structure makes more difficult in regartds of:
- API specification description -> this is not a problem but a cocnern for the OMeta / openaspi spec generator. We may need to check if this complex data structure will be modelled proeprly.
- Housekeeping the complex resource when the SBVR data model changes (owner of entity is deleted, moved.


I wonder if we don't want to make pineJS have some 'standard' SBVR resources that can have standard hooks on the resource. When I need a webresource I would add it to the parent-resource and let tghe webresource resource handle the work (fileStorage) in a hook?

Reply:
Hi @fisehara We went with this option of the complex attribute, instead of an standalone term, mainly to keep it simple in terms of permissions and to provide a one-step solution. But it is a bit controversial.
I'm happy to explore other alternatives and implement them so that we can compare.

Is there a way I can check the 1st issue you mention? Do you have a project that generates openapi that I can test a WebResource on?

About the second issue, I'm not sure I understand the case. What would be the owner of the entity?

Reply:
@fisehara When you mention:
> if we don't want to make pineJS have some 'standard' SBVR resources that can have standard hooks on the resource.

Does this still refer to use of a complex structure like above? 

The reason I ask is - we initially started with the idea of simple object and store related information with the owner of entity. But I think with while moving to a more generic solution this moved to a complex type which can have its own metadata.

Reply:
@nitishagar Thanks for the clarification. I'm fine with the complex-type. I wasn't fully aware of the reasoning to use it in favour to a sbvr resource.